src/Common/src/Interop/Unix/libcurl/Interop.SafeCurlHandle.cs

```diff
@@ -56,11 +96,175 @@ public static void DisposeAndClearHandle(ref SafeCurlMultiHandle curlHandle)
                 }
             }
 
+            internal void PollFds(List<libc.pollfd> readyFds)
+            {
+                int count;
+                libc.pollfd[] pollFds;
+
+                readyFds.Clear();
+
+                lock (this)
+                {
+                    // TODO: Avoid the allocation when count is in 100s
+                    count = _fdSet.Count + 1;
+                    pollFds = new libc.pollfd[count];
+
+                    // Always include special fd in the poll set
+                    pollFds[0].fd = _specialFds[libc.ReadEndOfPipe];
+                    pollFds[0].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                    int i = 1;
+                    foreach (int fd in _fdSet)
+                    {
+                        pollFds[i].fd = fd;
+                        pollFds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                        i++;
+                    }
+                }
+
+                unsafe
+                {
+                    fixed (libc.pollfd* fds = pollFds)
+                    {
+                        int numFds = libc.poll(fds, (uint)count, -1);
+                        lock (this)
+                        {
+                            if (0 == _fdCount)
+                            {
+                                return;
+                            }
+                        }
+                        if (numFds <= 0)
+                        {
+                            Debug.Assert(numFds != 0); // Since timeout is infinite
+                            // ignore errors
+                            return;
+                        }
+
+                        for (int i = 0, j = 0; i < count && j < numFds; i++)
+                        {
+                            if (fds[i].revents == 0)
+                            {
+                                continue;
+                            }
+                            if (i == 0)
```

Comment:
I still think it'd be easier to understand this logic if the 0 case was separated out to before the loop.  But I'll stop mentioning it.